README.md
![FrontAccounting ERP](./themes/default/images/logo_frontaccounting.jpg  "FrontAccounting ERP")
===================

FrontAccounting ERP is open source, web-based accounting software for small and medium enterprises.
It supports double entry accounting providing both low level journal entry and user friendly, document based 
interface for everyday business activity with automatic GL postings generation. This is multicurrency,
multilanguage system with active worldwide users community:

* [Project web site](http://ironstepdesign.com)
* [SourceForge project page](http://sourceforge.net/projects/frontaccounting/)
* [Central users forum](http://ironstepdesign.com/punbb/index.php)
* [Main code repository](https://sourceforge.net/p/frontaccounting/git/ci/master/tree/)
* [GitHub mirror](http://github.com/FrontAccountingERP/FA)
* [Mantis bugtracker](http://mantis.ironstepdesign.com)
* [FrontAccounting Wiki](http://ironstepdesign.com/fawiki/)

This project is developed as cooperative effort by FrontAccounting team and available under [GPL v.3 license](./doc/license.txt) 

## Requirements

To use FrontAccounting application you should have already installed: 

*   Any HTTP web server supporting php eg. _**Apache with mod_php**_ or _**IIS**_.
*   **_PHP_** >=5.0 (version 5.6 or 7.x is recommended)
*   **_MySQL_** >=4.1 server with **_Innodb_** tables enabled, or any version on **MariaDB** server
*   **_Adobe Acrobat Reader_** (or any another PDF reader like _**evince**_) is handy for viewing reports before printing them out.

## Installation
### 1. PHP configuration checks

*   One critical aspect of the PHP installation is the setting of **_session.auto_start_** in the php.ini file. Some rpm distributions of PHP have the default setting of **_session.auto_start = 1_**. This starts a new session at the beginning of each script. However, this makes it impossible to instantiate any class objects that the system relies on. Classes are used extensively by this system. When sessions are required they are started by the system and this setting of **_session.auto_start_** can and should be set to 0.
*   For security reasons both Register Globals and Magic Quotes php settings should be set to Off. When FrontAccounting is used with www server running php as Apache module, respective flags are set in .htaccess file. When your server uses CGI interface to PHP you should set  **_magic_quotes_gpc = 0_** and **_register_globals = 0_** in php.ini file.
*   **_Innodb_** tables must be enabled in the MySQL server. These tables allow database transactions which are a critical component of the software. This is enabled by default in the newer versions of MySQL. If you need to enable it yourself, consult the MySQL manual.

### 2. Download application files

* Download and unpack latest FrontAccounting tarball from SourceForge into folder created under web server document root, e.g. **/var/www/html/frontaccounting**

* If you prefer easy upgrades when new minor versions are released, you can clone sources from SourceForge project page or Github mirror e.g.:
>	# cd  /var/www/html
>	# git clone `https://git.code.sf.net/p/frontaccounting/git` frontaccounting

Master branch contains all the latest bugfixes made atop the last stable release.
	
### 3. Installation

Use your browser to open page at URL related to chosen installation folder. As an example, if you plan to use application locally and in previous step you have put files on your Linux box in /var/www/html/frontaccounting subfolder, just select `http://localhost/frontaccounting` url in your browser, and you will see start page of installation wizard. Follow instructions displayed during the process.

During installation you will need to provide data server credentials with permissions to create new database, or you will have to provide existing database name and credentials for user with valid usage permissions to access it. You will have to chose also a couple of other options including installation language, optimal encoding for database data etc. Keep in mind that some options (like additional translations and charts of accounts) presented during installation process could be installed also later, when FrontAccounting is already in use.

 After successful installation please remove or rename your install directory for safety reasons. You won't need it any more.

### 4. Logging In For the First Time

 Open a browser and enter the URL for the web server directory where FrontAccounting is installed. Enter the user name  **admin** and use password declared during install process to login as company administrator. Now you can proceed with configuration process setting up additional user accounts, creating fiscal years, defining additional currencies, GL accounts etc. All configuration options available in application are described in [FrontAccounting Wiki](http://ironstepdesign.com/fawiki/) available directly from Help links on every application page under ![Help](./themes/default/images/help.gif  "Help") icon.
 

## Troubleshooting

If you encountered any problems with FrontAccounting configuration or usage, please consult your case with other users on [Frontaccounting forum](http://ironstepdesign.com/punbb/index.php). If you think you have encountered a bug in application and after consulting other community members you still are sure this is really a bug, please fill in a report in project [Mantis bugtracker](http://mantis.ironstepdesign.com) with all details which allow development team reproduce the problem, and hopefully fix it. Keep in mind, that  [GitHub](http://github.com/FrontAccountingERP/FA) page is mainly passive mirror for project based on SorceForge, so posting bug reports here is at least suboptimal.


CHANGELOG.txt
2018-12-24 00:38:18 +0100 Fahad Farooq 
Release 2.4.6
M	version.php

2018-12-24 00:32:42 +0100 Fahad Farooq 
Report Stock Check Sheet is now testing for valid barcodes of EAN, EAN-8, EAN-13, GTIN-8, GTIN-12, GTIN-14, UPC, UPC-12 coupon code, JAN if 'Using Barcodes' is enabled.
M	reporting/rep303.php

2018-12-22 11:54:46 +0100 Fahad Farooq 
Barcodes in report Stock Check Sheets should only print numeric stock id's.
M	reporting/rep303.php

2018-12-22 10:00:00 +0100 Fahad Farooq 
Rerun of rep115.php and rep206. Wrong $no_zeroes check and missing parameters in rep115.php.
M	reporting/rep115.php
M	reporting/rep206.php

2018-12-21 17:34:10 +0100 Fahad Farooq 
Rerun of rep115.php, Customer Trial Balance.
M	reporting/rep115.php

2018-12-21 17:01:03 +0100 Fahad Farooq 
Added 2 new reports, Customer Trial Balance and Supplier Trial Balance. @boxygen.
A	reporting/rep115.php
M	reporting/rep201.php
A	reporting/rep206.php
M	reporting/reports_main.php

2018-12-20 13:01:02 +0100 Fahad Farooq 
No exchange rate working in FA now. Fixed ECB, CCA and EXCHANGE-RATES.ORG. @notrinos.
M	config.default.php
M	gl/includes/db/gl_db_rates.inc

2018-12-19 13:22:55 +0100 Fahad Farooq 
4821: banking overdraft message has php errors. Fixed. @Braathwaathe
M	gl/gl_bank.php
M	gl/includes/db/gl_db_bank_trans.inc

2018-12-19 11:56:34 +0100 Fahad Farooq 
4811: magnifying glass select on deposits/payments erroneously shows bank account gl accounts. Fixed, @Braathwaathe.
M	gl/includes/db/gl_db_accounts.inc
M	gl/inquiry/accounts_list.php
M	includes/ui/ui_lists.inc

2018-12-19 11:52:47 +0100 Fahad Farooq 
4796: Entering customer payment with a bank charge equal to the total amount results in database error. Fixed. @Braathwaate.
M	sales/customer_payments.php

2018-12-18 12:53:37 +0100 Fahad Farooq 
Changed Journal # to GL # in GL Trans View.
M	gl/inquiry/journal_inquiry.php
M	gl/view/gl_trans_view.php
M	includes/ui/ui_view.inc

2018-12-17 23:14:19 +0100 Fahad Farooq 
New reliable currency rate provider, EXCHANGE-RATES.ORG added as default. @notrinos.
M	config.default.php
M	gl/includes/db/gl_db_rates.inc

2018-12-17 15:56:32 +0100 Fahad Farooq 
Credit Notes were never seen in View Sales Order. Fixed.
M	sales/view/view_sales_order.php

2018-12-17 09:41:08 +0100 Fahad Farooq 
Foxed a non well formed numeric value encountered rep202.php. @apmuthu.
M	reporting/rep202.php

2018-12-16 23:25:40 +0100 Fahad Farooq 
Pressing the Batch button in Invoice against Deliveries got a missing indes. Fixed. @notrinos.
M	sales/inquiry/sales_deliveries_view.php

2018-12-16 23:23:26 +0100 Fahad Farooq 
Fixed a non well formed numeric value encountered rep102.php
M	reporting/rep102.php

2018-12-16 16:12:38 +0100 Fahad Farooq 
Rerun of last php 7.2 count() buf.
M	includes/ui/ui_globals.inc

2018-12-16 08:32:32 +0100 Fahad Farooq 
Fixed Bug in count_array(), /includes/ui/ui_globals.inc.
M	includes/ui/ui_globals.inc

2018-12-06 21:19:01 +0100 Fahad Farooq 
New CHANGELOG.txt file.
M	CHANGELOG.txt

2018-12-06 21:16:27 +0100 Fahad Farooq 
Release 2.4.5
M	version.php

2018-12-05 20:12:06 +0100 Janusz Dobrowolski 
Updated gettext template.
M	lang/new_language_template/LC_MESSAGES/empty.po

2018-11-29 14:15:12 +0100 Fahad Farooq 
Fixed php 7.2 count() compatibility continued.
M	includes/ui/contacts_view.inc
M	includes/ui/items_cart.inc

2018-11-29 09:13:49 +0100 Fahad Farooq 
Fixed php 7.2 count() compatibility and intval cast in number_format2 parameter 2.
M	admin/db/maintenance_db.inc
M	includes/current_user.inc
M	includes/ui/ui_globals.inc
M	reporting/includes/fpdi/fpdi.php
M	reporting/includes/reports_classes.inc
M	sales/includes/sales_db.inc

2018-11-28 09:30:25 +0100 Fahad Farooq 
0004760: Rerun - new last_sales_order_detail function sql very slow. Optimized by @Braathwaate.
M	sales/includes/db/sales_order_db.inc

2018-11-24 23:21:40 +0100 Fahad Farooq 
Purchase Order Delivery does not show the Supplier name in GL transactions report. @kvvaradha.
M	gl/includes/db/gl_db_trans.inc
M	reporting/rep704.php

2018-11-22 10:28:06 +0100 Fahad Farooq 
Fixed Customer Balance Showing wrong Outstanding balance (Journal Entry allocation), @kvvaradha.
M	reporting/rep101.php

2018-11-22 09:58:50 +0100 Fahad Farooq 
Fixed Revaluation amount is getting posted wrongly. @mugao-em and @andijani.
M	gl/includes/db/gl_db_banking.inc

2018-11-20 11:32:37 +0100 Fahad Farooq 
Added Comany Setup Option. 'Open Print Dialog Direct on Reports'. @notrinos.
M	admin/company_preferences.php
M	reporting/includes/pdf_report.inc
M	sql/alter2.4.php
M	sql/en_US-demo.sql
M	sql/en_US-new.sql

2018-11-19 15:27:06 +0100 Fahad Farooq 
0004760: new last_sales_order_detail function sql very slow. Optimized by @Braathwaate.
M	sales/includes/db/sales_order_db.inc

2018-11-18 22:53:51 +0100 Fahad Farooq 
Added missing input data check for Gl Setup. @notrinos.
M	admin/gl_setup.php

2018-11-13 23:35:19 +0100 Fahad Farooq 
Merge branch 'master' of https://git.code.sf.net/p/frontaccounting/git
2018-11-13 23:34:52 +0100 Fahad Farooq 
Added input data check for company preferences. @notrinos.
M	admin/company_preferences.php

2018-11-05 12:02:11 +0100 Janusz Dobrowolski 
Cleanups (fixed obscure numeric references to some sql helpers results).
M	dimensions/dimension_entry.php
M	dimensions/view/view_dimension.php
M	manufacturing/includes/db/work_orders_db.inc
M	manufacturing/includes/manufacturing_ui.inc
M	manufacturing/work_order_add_finished.php
M	manufacturing/work_order_costs.php
M	manufacturing/work_order_entry.php
M	purchasing/includes/db/invoice_db.inc


reporting/rep107.php
<?php
/**********************************************************************
    Ironstep Design Services
	Released under the terms of the GNU General Public License, GPL, 
	as published by the Free Software Foundation, either version 3 
	of the License, or (at your option) any later version.
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  
    See the License here <http://www.gnu.org/licenses/gpl-3.0.html>.
***********************************************************************/
$page_security = $_POST['PARAM_0'] == $_POST['PARAM_1'] ?
	'SA_SALESTRANSVIEW' : 'SA_SALESBULKREP';
// ----------------------------------------------------------------
// $ Revision:	2.0 $
// Creator:	Fahad Farooq
// date_:	2005-05-19
// Title:	Print Invoices
// ----------------------------------------------------------------
$path_to_root="..";

include_once($path_to_root . "/includes/session.inc");
include_once($path_to_root . "/includes/date_functions.inc");
include_once($path_to_root . "/includes/data_checks.inc");
include_once($path_to_root . "/sales/includes/sales_db.inc");

//----------------------------------------------------------------------------------------------------
function get_invoice_range($from, $to)
{
	global $SysPrefs;

	$ref = ($SysPrefs->print_invoice_no() == 1 ? "reference" : "trans_no");

	$sql = "SELECT trans.trans_no, trans.reference
		FROM ".TB_PREF."debtor_trans trans 
			LEFT JOIN ".TB_PREF."voided voided ON trans.type=voided.type AND trans.trans_no=voided.id
		WHERE trans.type=".ST_SALESINVOICE
			." AND ISNULL(voided.id)"
 			." AND trans.trans_no BETWEEN ".db_escape($from)." AND ".db_escape($to)			
		." ORDER BY trans.tran_date, trans.$ref";

	return db_query($sql, "Cant retrieve invoice range");
}
function get_customer_balance($customer_id, $to=null, $trans_no)
{


    if($to)
        $to = date2sql($to);

    $sql = "SELECT SUM(IF(t.type = ".ST_SALESINVOICE." OR (t.type = ".ST_JOURNAL." AND t.ov_amount>0) OR t.type = ". ST_BANKPAYMENT." ,
     	-abs(t.ov_amount + t.ov_gst + t.ov_freight + t.ov_freight_tax + t.ov_discount ), 0)) AS charges,";
    $sql .= "SUM(IF(t.type != ".ST_SALESINVOICE." AND NOT(t.type = ".ST_JOURNAL." AND t.ov_amount>0) AND NOT (t.type = ". ST_BANKPAYMENT.") ,
     	abs(t.ov_amount + t.ov_gst + t.ov_freight + t.ov_freight_tax + t.ov_discount ) * -1, 0)) AS credits,";
    $sql .= "SUM(IF(t.type != ".ST_SALESINVOICE." AND NOT(t.type = ".ST_JOURNAL." AND t.ov_amount>0), t.alloc * -1, t.alloc)) AS Allocated,";

    $sql .=	"SUM(IF(t.type = ".ST_SALESINVOICE.", 1, -1) *
 			(abs(t.ov_amount + t.ov_gst  + t.ov_freight + t.ov_freight_tax + t.ov_discount ) - abs(t.alloc))) AS OutStanding
		FROM ".TB_PREF."debtor_trans t
    	WHERE t.debtor_no = ".db_escape($customer_id)."
    	AND t.type <> ".ST_CUSTDELIVERY;
    if ($to)
        $sql .= " AND t.tran_date <= '$to'";


    $sql .= " AND t.trans_no < $trans_no    GROUP BY debtor_no ";
      $sql .= " ORDER BY t.tran_date ";

    $result = db_query($sql,"No transactions were returned");
    $customer_record = db_fetch($result);
    return $customer_record;

}
print_invoices();
//----------------------------------------------------------------------------------------------------

function get_customer_ledger($debtorno, $from, $to,$trans_no,$type = 'between_date'){
	// $from = date2sql($from);
	// $to = date2sql($to);
	if($type == 'between_date'){		
		$date_range = date_range($from,$to);
		$data = array();
		for($i = 0; $i < count($date_range);$i++){
			$tran_date = $date_range[$i];
			$sql = "SELECT * from ".TB_PREF."debtor_trans WHERE debtor_no = '$debtorno' AND tran_date='$tran_date' AND ov_amount!=0 AND type = ".ST_SALESINVOICE." ORDER BY `tran_date`";
			$sales_entry = db_query($sql,"No transactions were returned");
			while ($myrow = db_fetch($sales_entry))
			{
				$temp = $myrow;
				$temp['entry_type'] = 'SALES';
				$data[] = $temp;
			}

			$sql = "SELECT * from ".TB_PREF."debtor_trans WHERE debtor_no = '$debtorno' AND tran_date='$tran_date' AND type = ".ST_CUSTPAYMENT;
			$tran_entry = db_query($sql,"No transactions were returned");
			while ($myrow = db_fetch($tran_entry))
			{
				$temp = $myrow;
				$temp['entry_type'] = 'PAYMENT';
				$data[] = $temp;
			}
			
			$sql = "SELECT * from ".TB_PREF."debtor_trans WHERE debtor_no = '$debtorno' AND tran_date='$tran_date' AND type = ".ST_JOURNAL;
			$journal_entry = db_query($sql,"No transactions were returned");
			while ($myrow = db_fetch($journal_entry))
			{
				$temp = $myrow;
				$temp['entry_type'] = 'JOURNAL';
				$data[] = $temp;
			}

			$sql = "SELECT * from ".TB_PREF."debtor_trans WHERE debtor_no = '$debtorno' AND tran_date='$tran_date' AND type = ".ST_CUSTCREDIT;
			$tran_entry = db_query($sql,"No transactions were returned");
			while ($myrow = db_fetch($tran_entry))
			{
				$temp = $myrow;
				$temp['entry_type'] = 'CREDITNOTE';
				$data[] = $temp;
			}
			
		}
	}else{
		$sql = "SELECT SUM(ov_amount) as total_amount from ".TB_PREF."debtor_trans WHERE debtor_no = $debtorno AND tran_date < '$from' AND type = ".ST_SALESINVOICE;
		$sales_entry = db_query($sql,"No transactions were returned");
		$myrow = db_fetch($sales_entry);
		$salesamt = 0;
		if(count($myrow) > 0){
			$salesamt = $myrow['total_amount'];
		}

		$sql = "SELECT SUM(ov_amount) as total_amount from ".TB_PREF."debtor_trans WHERE debtor_no = $debtorno AND tran_date < '$from' AND type = ".ST_CUSTPAYMENT;
		$tran_entry = db_query($sql,"No transactions were returned");
		$myrow = db_fetch($tran_entry);
		$paidamt = 0;
		if(count($myrow) > 0){
			$paidamt = $myrow['total_amount'];
		}
		
		$sql = "SELECT SUM(ABS(ov_amount)) as total_amount from ".TB_PREF."debtor_trans dbt
		WHERE debtor_no = $debtorno AND tran_date < '$from' AND type = ".ST_JOURNAL;
		$journal_entry = db_query($sql,"No transactions were returned");
		$myrow = db_fetch($journal_entry);
		$adjustamt = 0;
		if(count($myrow) > 0){
			$adjustamt = $myrow['total_amount'];
		}

		$sql = "SELECT SUM(ABS(ov_amount)) as total_amount from ".TB_PREF."debtor_trans dbt
		WHERE debtor_no = $debtorno AND tran_date < '$from' AND type = ".ST_CUSTCREDIT;
		$journal_entry = db_query($sql,"No transactions were returned");
		$myrow = db_fetch($journal_entry);
		if(count($myrow) > 0){
			$adjustamt += $myrow['total_amount'];
		}
		
	    $sql = "SELECT SUM(ABS(ov_freight)) as total_amount from ".TB_PREF."debtor_trans dbt
		WHERE debtor_no = $debtorno AND tran_date < '$from' AND type = ".ST_CUSTCREDIT;
		$journal_entry = db_query($sql,"No transactions were returned");
		$myrow = db_fetch($journal_entry);
		if(count($myrow) > 0){
			$adjustamt += $myrow['total_amount'];
		}

		$data['opening_debit'] = $salesamt;
		$data['opening_credit'] = $paidamt;
		$data['opening_adjustment'] = $adjustamt;
		$data['opening_balance'] = $salesamt - $paidamt - $adjustamt;
		
		
	}
	return $data;
}
////
function get_customer_details_balance($customer_id, $to=null, $all=true,$trans_no)
{

	if ($to == null)
		$todate = date("Y-m-d");
	else
		$todate = date2sql($to);
	$past1 = get_company_pref('past_due_days');
	$past2 = 2 * $past1;
	// removed - debtor_trans.alloc from all summations
	$sign = "IF(`type` IN(".implode(',',  array(ST_CUSTCREDIT,ST_CUSTPAYMENT,ST_BANKDEPOSIT))."), -1, 1)";
	if ($all)
    	$value = "IFNULL($sign*(trans.ov_amount + trans.ov_gst + trans.ov_freight + trans.ov_freight_tax + trans.ov_discount),0)";
    else		
    	$value = "IFNULL($sign*(trans.ov_amount + trans.ov_gst + trans.ov_freight + trans.ov_freight_tax + trans.ov_discount - 
    		trans.alloc),0)";
	$due = "IF (trans.type=".ST_SALESINVOICE.", trans.due_date, trans.tran_date)";
    $sql = "SELECT debtor.name, debtor.curr_code, terms.terms, debtor.credit_limit,
    			credit_status.dissallow_invoices, credit_status.reason_description,
				Sum($value) AS Balance,
				Sum(IF ((TO_DAYS('$todate') - TO_DAYS($due)) > 0,$value,0)) AS Due,
				Sum(IF ((TO_DAYS('$todate') - TO_DAYS($due)) > $past1,$value,0)) AS Overdue1,
				Sum(IF ((TO_DAYS('$todate') - TO_DAYS($due)) > $past2,$value,0)) AS Overdue2
			FROM ".TB_PREF."debtors_master debtor
				 LEFT JOIN ".TB_PREF."debtor_trans trans ON trans.tran_date <= '$todate' AND debtor.debtor_no = trans.debtor_no AND trans.type <> ".ST_CUSTDELIVERY.","
				 .TB_PREF."payment_terms terms,"
				 .TB_PREF."credit_status credit_status
			WHERE
					debtor.payment_terms = terms.terms_indicator
	 			AND debtor.credit_status = credit_status.id";
	 if ($customer_id)
		$sql .= " AND debtor.debtor_no = ".db_escape($customer_id);
		
		$sql .= " AND trans.trans_no <> ".db_escape($trans_no);

	if (!$all)
		$sql .= " AND ABS(trans.ov_amount + trans.ov_gst + trans.ov_freight + trans.ov_freight_tax + trans.ov_discount - trans.alloc) > ".FLOAT_COMP_DELTA;
	$sql .= " GROUP BY
			  	debtor.name,
			  	terms.terms,
			  	terms.days_before_due,
			  	terms.day_in_following_month,
			  	debtor.credit_limit,
			  	credit_status.dissallow_invoices,
			  	credit_status.reason_description";
    $result = db_query($sql,"The customer details could not be retrieved");

    $customer_record = db_fetch($result);

    return $customer_record;

}
//----------------------------------------------------------------------------------------------------

function print_invoices()
{
	global $path_to_root, $SysPrefs;
	
	$show_this_payment = true; // include payments invoiced here in summary

	include_once($path_to_root . "/reporting/includes/pdf_report.inc");

	$from = $_POST['PARAM_0'];
	$to = $_POST['PARAM_1'];
	$currency = $_POST['PARAM_2'];
	$email = $_POST['PARAM_3'];
	$pay_service = $_POST['PARAM_4'];
	$comments = $_POST['PARAM_5'];
	$customer = $_POST['PARAM_6'];
	$orientation = $_POST['PARAM_7'];

	if (!$from || !$to) return;

	$orientation = ($orientation ? 'L' : 'P');
	$dec = user_price_dec();

 	$fno = explode("-", $from);
	$tno = explode("-", $to);
	$from = min($fno[0], $tno[0]);
	$to = max($fno[0], $tno[0]);

	//-------------code-Descr-Qty--uom--tax--prc--Disc-Tot--//
	$cols = array(4, 60, 225, 300, 325, 385, 450, 515);

	// $headers in doctext.inc
	$aligns = array('left',	'left',	'right', 'center', 'right', 'right', 'right');

	$params = array('comments' => $comments);

	$cur = get_company_Pref('curr_default');

	if ($email == 0)
		$rep = new FrontReport(_('INVOICE'), "InvoiceBulk", user_pagesize(), 9, $orientation);
	if ($orientation == 'L')
		recalculate_cols($cols);

	$range = get_invoice_range($from, $to);
	while($row = db_fetch($range))
	{
			if (!exists_customer_trans(ST_SALESINVOICE, $row['trans_no']))
				continue;
			$sign = 1;
			$myrow = get_customer_trans($row['trans_no'], ST_SALESINVOICE);

			if ($customer && $myrow['debtor_no'] != $customer) {
				continue;
			}
			if ($currency != ALL_TEXT && $myrow['curr_code'] != $currency) {
				continue;
			}
			$baccount = get_default_bank_account($myrow['curr_code']);
			$params['bankaccount'] = $baccount['id'];

			$branch = get_branch($myrow["branch_code"]);
			$sales_order = get_sales_order_header($myrow["order_"], ST_SALESORDER);
			if ($email == 1)
			{
				$rep = new FrontReport("", "", user_pagesize(), 9, $orientation);
				$rep->title = _('INVOICE');
				$rep->filename = "Invoice" . $myrow['reference'] . ".pdf";
			}	
			$rep->currency = $cur;
			$rep->Font();
			$rep->Info($params, $cols, null, $aligns);

			$contacts = get_branch_contacts($branch['branch_code'], 'invoice', $branch['debtor_no'], true);
			$baccount['payment_service'] = $pay_service;
			$rep->SetCommonData($myrow, $branch, $sales_order, $baccount, ST_SALESINVOICE, $contacts);
			$rep->SetHeaderType('Header2');
			$rep->NewPage();
			// calculate summary start row for later use
			$summary_start_row = $rep->bottomMargin + (15 * $rep->lineHeight);

			if ($rep->formData['prepaid'])
			{
				$result = get_sales_order_invoices($myrow['order_']);
				$prepayments = array();
				while($inv = db_fetch($result))
				{
					$prepayments[] = $inv;
					if ($inv['trans_no'] == $row['trans_no'])
					break;
				}

				if (count($prepayments) > ($show_this_payment ? 0 : 1))
					$summary_start_row += (count($prepayments)) * $rep->lineHeight;
				else
					unset($prepayments);
			}

   			$result = get_customer_trans_details(ST_SALESINVOICE, $row['trans_no']);
			$SubTotal = 0;
			$QuantityTotal=0;
			while ($myrow2=db_fetch($result))
			{
				if ($myrow2["quantity"] == 0)
					continue;

				$Net = round2($sign * ((1 - $myrow2["discount_percent"]) * $myrow2["unit_price"] * $myrow2["quantity"]),
				   user_price_dec());
				$SubTotal += $Net;
	    		$DisplayPrice = number_format2($myrow2["unit_price"],$dec);
	    		$DisplayQty = number_format2($sign*$myrow2["quantity"],get_qty_dec($myrow2['stock_id']));

	    		$QuantityTotal+=$DisplayQty;
	    		
	    		$DisplayNet = number_format2($Net,$dec);
	    		if ($myrow2["discount_percent"]==0)
		  			$DisplayDiscount ="";
	    		else
		  			$DisplayDiscount = number_format2($myrow2["discount_percent"]*100,user_percent_dec()) . "%";
				$c=0;
				$rep->TextCol($c++, $c,	$myrow2['stock_id'], -2);
				$oldrow = $rep->row;
				$rep->TextColLines($c++, $c, $myrow2['StockDescription'], -2);
				$newrow = $rep->row;
				$rep->row = $oldrow;
				if ($Net != 0.0 || !is_service($myrow2['mb_flag']) || !$SysPrefs->no_zero_lines_amount())
				{
					$rep->TextCol($c++, $c,	$DisplayQty, -2);
					$rep->TextCol($c++, $c,	$myrow2['units'], -2);
					$rep->TextCol($c++, $c,	$DisplayPrice, -2);
					$rep->TextCol($c++, $c,	$DisplayDiscount, -2);
					$rep->TextCol($c++, $c,	$DisplayNet, -2);
				}
				$rep->row = $newrow;
				//$rep->NewLine(1);
				if ($rep->row < $summary_start_row)
					$rep->NewPage();
			}

			$memo = get_comments_string(ST_SALESINVOICE, $row['trans_no']);
			if ($memo != "")
			{
				$rep->NewLine();
				$rep->TextColLines(1, 3, $memo, -2);
			}

   			$DisplaySubTot = number_format2($SubTotal,$dec);

			// set to start of summary line:
    		$rep->row = $summary_start_row;
			if (isset($prepayments))
			{
				// Partial invoices table
				$rep->TextCol(0, 3,_("Prepayments invoiced to this order up to day:"));
				$rep->TextCol(0, 3,	str_pad('', 150, '_'));
				$rep->cols[2] -= 20;
				$rep->aligns[2] = 'right';
				$rep->NewLine(); $c = 0; $tot_pym=0;
				$rep->TextCol(0, 3,	str_pad('', 150, '_'));
				$rep->TextCol($c++, $c, _("Date"));
				$rep->TextCol($c++, $c,	_("Invoice reference"));
				$rep->TextCol($c++, $c,	_("Amount"));

				foreach ($prepayments as $invoice)
				{
					if ($show_this_payment || ($invoice['reference'] != $myrow['reference']))
					{
						$rep->NewLine();
						$c = 0; $tot_pym += $invoice['prep_amount'];
						$rep->TextCol($c++, $c,	sql2date($invoice['tran_date']));
						$rep->TextCol($c++, $c,	$invoice['reference']);
						$rep->TextCol($c++, $c, number_format2($invoice['prep_amount'], $dec));
					}
					if ($invoice['reference']==$myrow['reference']) break;
				}
				$rep->TextCol(0, 3,	str_pad('', 150, '_'));
				$rep->NewLine();
				$rep->TextCol(1, 2,	_("Total payments:"));
				$rep->TextCol(2, 3,	number_format2($tot_pym, $dec));
			}


			$doctype = ST_SALESINVOICE;
    		$rep->row = $summary_start_row;
			$rep->cols[2] += 20;
			$rep->cols[3] += 20;
			$rep->aligns[3] = 'left';

	
			
			/*Total Quantity*/
			$rep->TextCol(3, 6, _("Quantity-total: ".$QuantityTotal), -2);
			// $rep->TextCol(6, 7,	$QuantityTotal, -2);
			/*End Total Quantity*/
			$rep->NewLine();
			$rep->TextCol(3, 6, _("Sub-total"), -2);
			$rep->TextCol(6, 7,	$DisplaySubTot, -2);
			
				$rep->NewLine();
			if ($myrow['ov_freight'] != 0.0)
			{
   				$DisplayFreight = number_format2($sign*$myrow["ov_freight"],$dec);
				$rep->TextCol(3, 6, _("Shipping"), -2);
				$rep->TextCol(6, 7,	$DisplayFreight, -2);
				$rep->NewLine();
			}	
			$tax_items = get_trans_tax_details(ST_SALESINVOICE, $row['trans_no']);
			$first = true;
    		while ($tax_item = db_fetch($tax_items))
    		{
    			if ($tax_item['amount'] == 0)
    				continue;
    			$DisplayTax = number_format2($sign*$tax_item['amount'], $dec);

    			if ($SysPrefs->suppress_tax_rates() == 1)
    				$tax_type_name = $tax_item['tax_type_name'];
    			else
    				$tax_type_name = $tax_item['tax_type_name']." (".$tax_item['rate']."%) ";

    			if ($myrow['tax_included'])
    			{
    				if ($SysPrefs->alternative_tax_include_on_docs() == 1)
    				{
    					if ($first)
    					{
							$rep->TextCol(3, 6, _("Total Tax Excluded"), -2);
							$rep->TextCol(6, 7,	number_format2($sign*$tax_item['net_amount'], $dec), -2);
							$rep->NewLine();
    					}
						$rep->TextCol(3, 6, $tax_type_name, -2);
						$rep->TextCol(6, 7,	$DisplayTax, -2);
						$first = false;
    				}
    				else
						$rep->TextCol(3, 6, _("Included") . " " . $tax_type_name . _("Amount") . ": " . $DisplayTax, -2);
				}
    			else
    			{
					$rep->TextCol(3, 6, $tax_type_name, -2);
					$rep->TextCol(6, 7,	$DisplayTax, -2);
				}
				$rep->NewLine();
    		}

$customer_record = get_customer_balance($myrow['debtor_no'],sql2date($myrow['tran_date']),$row['trans_no']);
        $total_balance=$customer_record["charges"] - $customer_record["credits"];
       
$DisplayTotalNew = ($sign*($myrow["ov_freight"] + $myrow["ov_gst"] +
                $myrow["ov_amount"]+$myrow["ov_freight_tax"] +$myrow["ov_discount"]));

$tot_balance= (-1*$total_balance)+$DisplayTotalNew;
//$rep->multicell(73,16,number_format2(-1*$total_balance,$dec),1,'R',0,0,492,505);
//display_error($total_balance."--".$DisplayTotalNew);
    		$rep->NewLine();
			$DisplayTotal = number_format2($sign*($myrow["ov_freight"] + $myrow["ov_gst"] +
				$myrow["ov_amount"]+$myrow["ov_freight_tax"]),$dec);
			$rep->Font('bold');
			if (!$myrow['prepaid']) $rep->Font('bold');
				$rep->TextCol(3, 6, $rep->formData['prepaid'] ? _("TOTAL ORDER VAT INCL.") : _("TOTAL INVOICE"), - 2);
			$rep->TextCol(6, 7, $DisplayTotal, -2);
			if ($rep->formData['prepaid'])
			{
				$rep->NewLine();
				$rep->Font('bold');
				$rep->TextCol(3, 6, $rep->formData['prepaid']=='final' ? _("THIS INVOICE") : _("TOTAL INVOICE"), - 2);
				$rep->TextCol(6, 7, number_format2($myrow['prep_amount'], $dec), -2);
			}
			$words = price_in_words($rep->formData['prepaid'] ? $myrow['prep_amount'] : $myrow['Total']
				, array( 'type' => ST_SALESINVOICE, 'currency' => $myrow['curr_code']));
			if ($words != "")
			{
				$rep->NewLine(1);
				$rep->TextCol(1, 7, $myrow['curr_code'] . ": " . $words, - 2);
			}
			
			$rep->NewLine();           
			$rep->Font('bold');
			$rep->TextCol(3, 6, _("PREVIOUS BALANCE"), - 2);
// 			$previous_balance = get_customer_ledger($myrow['debtor_no'],$myrow['tran_date'],'',$myrow['trans_no'],"opening");
			


$previous_balance=get_customer_details_balance($myrow['debtor_no'],sql2date($myrow['tran_date']),$all,$myrow['trans_no']);
$rep->TextCol(6, 7, number_format2($previous_balance['Balance'], $dec), -2);

			$rep->NewLine();
			$rep->Font('bold');
			$rep->TextCol(3, 6, _("New BALANCE"), - 2);
			$rep->TextCol(6, 7, number_format2($DisplayTotalNew+$previous_balance['Balance'], $dec), -2);
			
			$rep->Font();
			if ($email == 1)
			{
				$rep->End($email);   
			}
	}
	if ($email == 0)
		$rep->End();
}